packages/ckeditor5-engine/tests/view/renderer.js

```diff
@@ -4472,6 +4507,78 @@ describe( 'Renderer', () => {
 			expect( checkMappings() ).to.be.true;
 		} );
 
+		it( 'should properly render if text is changed and text and element is inserted into same node', () => {
+			setViewData( view,
+				'<container:p>foo<attribute:strong>123</attribute:strong>456</container:p>'
+			);
+
+			// Render it to DOM to create initial DOM <-> view mappings.
+			view.forceRender();
+			cleanObserver( observer );
+
+			// Modify the view.
+			view.change( writer => {
+				writer.insert(
+					writer.createPositionAfter( viewRoot.getChild( 0 ).getChild( 0 ) ),
+					parse( 'bar<attribute:strong>abc</attribute:strong>' )
+				);
+			} );
+
+			expect( getViewData( view ) ).to.equal( '<p>foobar<strong>abc123</strong>456</p>' );
+
+			// Re-render changes in view to DOM.
+			view.forceRender();
+
+			// Check if DOM is rendered correctly.
+			expect( normalizeHtml( domRoot.innerHTML ) ).to.equal( '<p>foobar<strong>abc123</strong>456</p>' );
+			expect( checkMappings() ).to.be.true;
+
+			expect( getMutationStats( observer.takeRecords() ) ).to.deep.equal( [
+				'updated text: "foo" to "foobar"',
+				'updated text: "123" to "abc123"'
+			] );
+		} );
+
+		it( 'should properly render if text is replaced by an element and following text', () => {
+			setViewData( view,
+				'<container:p>foo<attribute:strong>123</attribute:strong>456</container:p>'
+			);
+
+			// Render it to DOM to create initial DOM <-> view mappings.
+			view.forceRender();
+			cleanObserver( observer );
+
+			// Modify the view.
+			view.change( writer => {
+				writer.remove( viewRoot.getChild( 0 ).getChild( 0 ) );
+				writer.insert(
+					writer.createPositionAt( viewRoot.getChild( 0 ), 0 ),
+					parse(
+						'<attribute:strong>abc</attribute:strong>' +
+						'bar' +
+						'<attribute:strong>xyz</attribute:strong>'
+					)
+				);
+			} );
+
+			expect( getViewData( view ) ).to.equal( '<p><strong>abc</strong>bar<strong>xyz123</strong>456</p>' );
+
+			// Re-render changes in view to DOM.
+			view.forceRender();
+
+			// Check if DOM is rendered correctly.
+			expect( normalizeHtml( domRoot.innerHTML ) ).to.equal( '<p><strong>abc</strong>bar<strong>xyz123</strong>456</p>' );
+			expect( checkMappings() ).to.be.true;
+
+			expect( getMutationStats( observer.takeRecords() ) ).to.deep.equal( [
+				'added: [], removed: [ text: "foo" ]',
+				'added: [ text: "bar" ], removed: []',
+				'added: [ <strong> ], removed: []',
+				'updated text: "123" to "abc"',
+				'updated text: "abc123" to "abc"'
```

Comment:
I just added some comments to that test. What happens, in this case, is a text node update in 2 steps (insert `"abc"` to the existing text node and then remove `"123"` from that node). This log from the mutation observer is not perfect because it can't display the value after the change, only the final value of the modified node.